std/assembly/util/number.ts

```diff
@@ -203,118 +245,219 @@ function utoa64_lut(buffer: usize, num: u64, offset: usize): void {
   } while (num);
 }
 
+function utoa_hex_simple<T extends number>(buffer: usize, num: T, offset: usize): void {
+  do {
+    let d = changetype<T>(num) & 0x0F | CharCode._0;
+    d += select<T>(<T>0x27, <T>0, d > <T>CharCode._9);
+    offset--;
+    store<u16>(buffer + (offset << 1), d);
+    // @ts-ignore: type
+    num >>= 4;
+  } while (num);
+}
+
 // @ts-ignore: decorator
 @inline
-export function utoa32_core(buffer: usize, num: u32, offset: u32): void {
+export function utoa32_dec_core(buffer: usize, num: u32, offset: usize): void {
   if (ASC_SHRINK_LEVEL >= 1) {
-    utoa_simple(buffer, num, offset);
+    utoa_dec_simple<u32>(buffer, num, offset);
   } else {
-    utoa32_lut(buffer, num, offset);
+    utoa32_dec_lut(buffer, num, offset);
   }
 }
 
 // @ts-ignore: decorator
 @inline
-export function utoa64_core(buffer: usize, num: u64, offset: u32): void {
+export function utoa32_hex_core(buffer: usize, num: u32, offset: usize): void {
   if (ASC_SHRINK_LEVEL >= 1) {
-    utoa_simple(buffer, num, offset);
+    utoa_hex_simple<u32>(buffer, num, offset);
   } else {
-    utoa64_lut(buffer, num, offset);
+    utoa_hex_lut(buffer, num, offset);
   }
 }
 
-export function utoa32(value: u32): String {
-  if (!value) return "0";
+// @ts-ignore: decorator
+@inline
+export function utoa64_dec_core(buffer: usize, num: u64, offset: usize): void {
+  if (ASC_SHRINK_LEVEL >= 1) {
+    utoa_dec_simple<u64>(buffer, num, offset);
+  } else {
+    utoa64_dec_lut(buffer, num, offset);
+  }
+}
 
-  var decimals = decimalCount32(value);
-  var out = __alloc(decimals << 1, idof<String>());
+// @ts-ignore: decorator
+@inline
+export function utoa64_hex_core(buffer: usize, num: u64, offset: usize): void {
+  if (ASC_SHRINK_LEVEL >= 1) {
+    utoa_hex_simple<u64>(buffer, num, offset);
+  } else {
+    utoa_hex_lut(buffer, num, offset);
+  }
+}
 
-  utoa32_core(out, value, decimals);
+export function utoa64_any_core(buffer: usize, num: u64, offset: usize, radix: i32): void {
+  const lut = changetype<usize>(ANY_DIGITS);
+  var base = u64(radix);
+  if ((radix & (radix - 1)) == 0) { // for radix which pow of two
+    let shift = u64(ctz(radix) & 7);
+    let mask = base - 1;
+    do {
+      offset--;
+      store<u16>(buffer + (offset << 1), load<u16>(lut + (usize(num & mask) << 1)));
+      num >>= shift;
+    } while (num);
+  } else {
+    do {
+      offset--;
+      let q = num / base;
+      store<u16>(buffer + (offset << 1), load<u16>(lut + (usize(num - q * base) << 1)));
+      num = q;
+    } while (num);
+  }
+}
+
+export function utoa32(value: u32, radix: i32): String {
+  if (radix < 2 || radix > 36) {
+    throw new RangeError("toString() radix argument must be between 2 and 36");
+  }
+  if (!value) return "0";
+  var out: usize = 0;
+
+  if (radix == 10) {
+    let decimals = decimalCount32(value);
+    out = __alloc(decimals << 1, idof<String>());
+    utoa32_dec_core(out, value, decimals);
+  } else if (radix == 16) {
```

Comment:
Would it make sense to also handle other common cases, like radix 2 or 8?

Reply:
I don't think so. This bases have rarely usage and usually for debugging / tracing proposes so I don't think we should do special optimizations for them. Btw `utoa64_any_core` already has fast path for pow of two bases like (2, 8, ..., 32) which don't use shifts instead division so it's pretty enough.